File: app/src/main/java/org/puder/trs80/EditConfigurationFragment.java
/*
 * Copyright 2012-2013, Arno Puder
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.puder.trs80;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.preference.Preference;
import android.preference.Preference.OnPreferenceChangeListener;
import android.preference.PreferenceFragment;
import android.preference.PreferenceManager;
import android.util.Log;

import com.google.common.base.Optional;

import org.puder.trs80.browser.FileBrowserActivity;
import org.puder.trs80.configuration.ConfigurationPersistence;
import org.puder.trs80.configuration.ConfigurationPersistence.PreferenceFinder;
import org.puder.trs80.configuration.ConfigurationPersistence.PreferenceProvider;

public class EditConfigurationFragment extends PreferenceFragment implements
        OnPreferenceChangeListener {

    private Handler           handler;

    private boolean           configurationWasEdited;

    private ConfigurationPersistence  configPersitence;
    private PreferenceFinder  prefFinder;
    private Preference        model;
    private Preference        name;
    private Preference        cassette;
    private Preference        disk1;
    private Preference        disk2;
    private Preference        disk3;
    private Preference        disk4;
    private Preference        characterColor;
    private Preference        keyboardPortrait;
    private Preference        keyboardLandscape;

    private String      defaultCassetteSummary;
    private String      defaultDisk1Summary;
    private String      defaultDisk2Summary;
    private String      defaultDisk3Summary;
    private String      defaultDisk4Summary;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        configurationWasEdited = false;
        handler = new Handler();
        Intent i = getActivity().getIntent();
        int configId = i.getExtras().getInt("CONFIG_ID");
        configPersitence = ConfigurationPersistence.forIdAndManager(
                configId, getPreferenceManager());
        prefFinder = configPersitence.forPreferenceProvider(new PreferenceProvider() {
            @Override
            public Preference findPreference(String name) {
                return EditConfigurationFragment.this.findPreference(name);
            }
        });
        addPreferencesFromResource(R.xml.configuration);
        name = prefFinder.forName();
        name.setOnPreferenceChangeListener(this);
        Preference.OnPreferenceClickListener listener = new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference pref) {
                String key = pref.getKey();
                int requestCode;
                if (key.equals("conf_cassette")) {
                    requestCode = 0;
                } else {
                    requestCode = Integer.parseInt(key.substring(key.length() - 1));
                }
                Intent intent = new Intent(getActivity(), FileBrowserActivity.class);
                startActivityForResult(intent, requestCode);
                return true;
            }
        };

        cassette = prefFinder.forCasette();
        cassette.setOnPreferenceChangeListener(this);
        cassette.setOnPreferenceClickListener(listener);
        defaultCassetteSummary = cassette.getSummary().toString();

        disk1 = prefFinder.forDisk1();
        disk1.setOnPreferenceChangeListener(this);
        disk1.setOnPreferenceClickListener(listener);
        defaultDisk1Summary = disk1.getSummary().toString();

        disk2 = prefFinder.forDisk2();
        disk2.setOnPreferenceChangeListener(this);
        disk2.setOnPreferenceClickListener(listener);
        defaultDisk2Summary = disk2.getSummary().toString();

        disk3 = prefFinder.forDisk3();
        disk3.setOnPreferenceChangeListener(this);
        disk3.setOnPreferenceClickListener(listener);
        defaultDisk3Summary = disk3.getSummary().toString();

        disk4 = prefFinder.forDisk4();
        disk4.setOnPreferenceChangeListener(this);
        disk4.setOnPreferenceClickListener(listener);
        defaultDisk4Summary = disk4.getSummary().toString();

        model = prefFinder.forModel();
        model.setOnPreferenceChangeListener(this);

        characterColor = prefFinder.forCharacterColor();
        characterColor.setOnPreferenceChangeListener(this);

        keyboardPortrait = prefFinder.forKeyboardPortrait();
        keyboardPortrait.setOnPreferenceChangeListener(this);

        keyboardLandscape = prefFinder.forKeyboardLandscape();
        keyboardLandscape.setOnPreferenceChangeListener(this);

        updateSummaries();
    }

    private void updateSummaries() {
        Optional<String> valOpt = configPersitence.getName();
        if (valOpt.isPresent()) {
            name.setSummary(valOpt.get());
        }
        valOpt = configPersitence.getModel();
        if (valOpt.isPresent()) {
            String val = valOpt.get();
            if (val.equals("1")) {
                val = "I";
            } else if (val.equals("3")) {
                val = "III";
            } else if (val.equals("5")) {
                val = "4P";
            }
            model.setSummary("Model " + val);
        }

        // Cassette
        valOpt = configPersitence.getCasettePath();
        cassette.setSummary(valOpt.or(defaultCassetteSummary));

        // Disk 1
        valOpt = configPersitence.getDiskPath(0);
        disk1.setSummary(valOpt.or(defaultDisk1Summary));

        // Disk 2
        valOpt = configPersitence.getDiskPath(1);
        disk2.setSummary(valOpt.or(defaultDisk2Summary));

        // Disk 3
        valOpt = configPersitence.getDiskPath(2);
        disk3.setSummary(valOpt.or(defaultDisk3Summary));

        // Disk 4
        valOpt = configPersitence.getDiskPath(3);
        disk4.setSummary(valOpt.or(defaultDisk4Summary));

        // Character color
        setCharacterColorSummary(configPersitence.getCharacterColor(0));

        // Keyboard portrait
        setKeyboardSummary(keyboardPortrait, configPersitence.getKeyboardLayoutPortrait());

        // Keyboard landscape
        setKeyboardSummary(keyboardLandscape, configPersitence.getKeyboardLayoutLandscape());
    }

    private void setCharacterColorSummary(int val) {
        switch (val) {
            case 0:
                characterColor.setSummary(this.getString(R.string.green));
                return;
            case 1:
                characterColor.setSummary(this.getString(R.string.white));
                return;
            default:
                return;
        }
    }

    private void setKeyboardSummary(Preference pref, int val) {
        switch (val) {
            case 0:
                pref.setSummary(this.getString(R.string.keyboard_original));
                break;
            case 1:
                pref.setSummary(this.getString(R.string.keyboard_compact));
                break;
            case 2:
                pref.setSummary(this.getString(R.string.keyboard_joystick));
                break;
            case 3:
                pref.setSummary(R.string.keyboard_game_controller);
            case 4:
                pref.setSummary(this.getString(R.string.keyboard_tilt));
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (resultCode == Activity.RESULT_OK) {
            configurationWasEdited = true;
            String newValue = data.getStringExtra("PATH");
            switch (requestCode){
                case 0:
                    configPersitence.setCasettePath(newValue);
                    break;
                case 1:
                case 2:
                case 3:
                case 4:
                    configPersitence.setDiskPath(requestCode - 1, newValue);

            }
            updateSummaries();
        }
    }

    @Override
    public boolean onPreferenceChange(Preference preference, Object newValue) {
        configurationWasEdited = true;
        /*
         * When we get to this point, the preferences have not yet been updated
         * yet. For this reason updateSummaries() is called via a handler to
         * ensure the preferences have been updated.
         */
        handler.post(new Runnable() {

            @Override
            public void run() {
                updateSummaries();
            }
        });
        return true;
    }

    public boolean configurationWasEdited() {
        return configurationWasEdited;
    }
}

File: app/src/main/java/org/puder/trs80/EditConfigurationFragment.kt
/*
 * Copyright 2012-2013, Arno Puder
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.puder.trs80

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.View
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import org.puder.trs80.browser.FileBrowserActivity
import org.puder.trs80.configuration.ConfigurationManager
import org.puder.trs80.configuration.ConfigurationPersistence
import org.puder.trs80.shared.storage.StorageKeys
import org.puder.trs80.storage.AppStorage
import org.puder.trs80.storage.SettingsPreferenceDataStore
import java.io.File
import java.io.IOException

private const val TAG = "EditConfFragment"

/** Intent extras shared with [EditConfigurationActivity]'s caller and the file browser. */
private const val EXTRA_CONFIG_ID = "CONFIG_ID"
private const val EXTRA_DIR = "DIR"
private const val EXTRA_PATH = "PATH"

/** Key of the cassette preference, which is the one file preference not named `conf_diskN`. */
private const val KEY_CASSETTE = "conf_cassette"

/** Request code used when browsing for a cassette image; disk drive n uses n + 1. */
private const val REQUEST_CASSETTE = 0

/** The number of disk drives a configuration has. */
private const val DISK_COUNT = 4

/**
 * Edits a single configuration. Every change is written straight into the configuration's
 * preferences, so [configurationWasEdited] is what tells the caller whether it has to restore
 * its backup.
 */
class EditConfigurationFragment : PreferenceFragmentCompat() {

    /** Whether the user changed anything since this fragment was created. */
    var configurationWasEdited = false
        private set

    private val handler = Handler(Looper.getMainLooper())

    private val changeListener = Preference.OnPreferenceChangeListener { _, _ ->
        configurationWasEdited = true
        // The preferences have not been written yet at this point, so refresh the summaries
        // via the handler, once the pending change has been applied.
        handler.post { updateSummaries() }
        true
    }

    private lateinit var configPersistence: ConfigurationPersistence
    private lateinit var namePref: Preference
    private lateinit var modelPref: Preference
    private lateinit var characterColorPref: Preference
    private lateinit var keyboardPortraitPref: Preference
    private lateinit var keyboardLandscapePref: Preference
    private lateinit var cassette: MediaPreference
    private lateinit var disks: List<MediaPreference>

    // startActivityForResult; the file browser it drives is replaced by a document picker
    // rather than migrated to the Activity Result API.
    @Suppress("DEPRECATION")
    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        val configId = ActivityHelper.getIntExtra(requireActivity().intent, EXTRA_CONFIG_ID)
        if (configId == null) {
            Log.w(TAG, "Cannot get CONFIG_ID. Finishing activity.")
            requireActivity().finish()
            return
        }
        // Point the screen at this configuration's namespace in the shared store,
        // which has to happen before the hierarchy below is inflated so that the
        // preferences read their initial values from the right place.
        preferenceManager.preferenceDataStore = SettingsPreferenceDataStore(
            AppStorage.get().settings,
            StorageKeys.configurationPrefix(configId),
        )
        configPersistence = ConfigurationPersistence.forId(configId)
        val prefFinder = configPersistence.forPreferenceProvider { name ->
            requireNotNull(findPreference<Preference>(name)) { "No preference named '$name'." }
        }
        setPreferencesFromResource(R.xml.configuration, rootKey)

        namePref = prefFinder.forName()
        namePref.onPreferenceChangeListener = changeListener

        val clickListener = Preference.OnPreferenceClickListener { pref ->
            val key = pref.key
            val requestCode: Int
            val currentPath: String?
            if (key == KEY_CASSETTE) {
                requestCode = REQUEST_CASSETTE
                currentPath = configPersistence.casettePath
            } else {
                requestCode = key.last().digitToInt()
                currentPath = configPersistence.getDiskPath(requestCode - 1)
            }

            // Start out in the directory of the currently selected image, or in the
            // configuration's own directory while nothing is selected.
            val dir = if (currentPath != null) {
                File(currentPath).parent
            } else {
                try {
                    ConfigurationManager.get(requireContext())
                        .getEmulatorState(configId).basePath
                } catch (e: IOException) {
                    Log.w(TAG, "Could not get ConfigurationManager. Finishing activity.", e)
                    requireActivity().finish()
                    null
                }
            }

            val intent = Intent(requireContext(), FileBrowserActivity::class.java)
            if (dir != null) {
                intent.putExtra(EXTRA_DIR, dir)
            }
            startActivityForResult(intent, requestCode)
            true
        }

        cassette = mediaPreference(prefFinder.forCasette(), clickListener)
        disks = (0 until DISK_COUNT)
            .map { mediaPreference(prefFinder.forDisk(it), clickListener) }

        modelPref = prefFinder.forModel()
        modelPref.onPreferenceChangeListener = changeListener

        characterColorPref = prefFinder.forCharacterColor()
        characterColorPref.onPreferenceChangeListener = changeListener

        keyboardPortraitPref = prefFinder.forKeyboardPortrait()
        keyboardPortraitPref.onPreferenceChangeListener = changeListener

        keyboardLandscapePref = prefFinder.forKeyboardLandscape()
        keyboardLandscapePref.onPreferenceChangeListener = changeListener

        updateSummaries()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        view.fitsSystemWindows = true
    }

    @Suppress("DEPRECATION", "OVERRIDE_DEPRECATION") // See onCreatePreferences.
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (resultCode != Activity.RESULT_OK) {
            return
        }
        configurationWasEdited = true
        // A null path means the user ejected the image, which clears the setting.
        val newValue = data?.getStringExtra(EXTRA_PATH)
        when (requestCode) {
            REQUEST_CASSETTE -> configPersistence.setCasettePath(newValue)
            in 1..DISK_COUNT -> configPersistence.setDiskPath(requestCode - 1, newValue)
        }
        updateSummaries()
    }

    private fun mediaPreference(
        pref: Preference,
        clickListener: Preference.OnPreferenceClickListener
    ): MediaPreference {
        pref.onPreferenceChangeListener = changeListener
        pref.onPreferenceClickListener = clickListener
        return MediaPreference(pref, pref.summary.toString())
    }

    private fun updateSummaries() {
        configPersistence.name?.let { namePref.summary = it }
        configPersistence.model?.let { modelPref.summary = "Model " + modelLabel(it) }

        cassette.preference.summary = configPersistence.casettePath ?: cassette.defaultSummary
        disks.forEachIndexed { drive, disk ->
            disk.preference.summary = configPersistence.getDiskPath(drive) ?: disk.defaultSummary
        }

        setCharacterColorSummary(configPersistence.getCharacterColor(0))
        setKeyboardSummary(keyboardPortraitPref, configPersistence.keyboardLayoutPortrait)
        setKeyboardSummary(keyboardLandscapePref, configPersistence.keyboardLayoutLandscape)
    }

    /** Maps a stored model value onto the way the model is spelled in the UI. */
    private fun modelLabel(model: String): String = when (model) {
        "1" -> "I"
        "3" -> "III"
        "5" -> "4P"
        else -> model
    }

    private fun setCharacterColorSummary(color: Int) {
        val summaryId = when (color) {
            0 -> R.string.green
            1 -> R.string.white
            else -> return
        }
        characterColorPref.setSummary(summaryId)
    }

    private fun setKeyboardSummary(pref: Preference, layout: Int) {
        val summaryId = when (layout) {
            0 -> R.string.keyboard_original
            1 -> R.string.keyboard_compact
            2 -> R.string.keyboard_joystick
            // The Java original fell through from the game controller case into the tilt case,
            // so layout 3 has always ended up labelled "tilt" as well. Kept as is; changing the
            // label is a behavioural fix that does not belong in this conversion.
            3, 4 -> R.string.keyboard_tilt
            else -> return
        }
        pref.setSummary(summaryId)
    }

    /** A file-valued preference plus the summary to fall back to while no file is set. */
    private class MediaPreference(val preference: Preference, val defaultSummary: String)
}
